dev/com.ibm.ws.opentracing.1.3/src/com/ibm/ws/opentracing/AppStateListener.java

```diff
@@ -0,0 +1,92 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.opentracing;
+
+import java.util.Dictionary;
+import java.util.Hashtable;
+import java.util.Map;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+import com.ibm.ws.container.service.app.deploy.ApplicationInfo;
+import com.ibm.ws.container.service.state.ApplicationStateListener;
+import com.ibm.ws.container.service.state.StateChangeException;
+import com.ibm.wsspi.kernel.service.utils.ConcurrentServiceReferenceSet;
+import com.ibm.wsspi.library.Library;
+import com.ibm.wsspi.library.LibraryChangeListener;
+
+@Component(service = { ApplicationStateListener.class }, configurationPolicy = ConfigurationPolicy.IGNORE, immediate = true, property = "library=unbound")
+public class AppStateListener implements ApplicationStateListener, LibraryChangeListener {
+
+    private static final String LIBRARY = "library";
+    private volatile ConcurrentServiceReferenceSet<Library> librarySet = new ConcurrentServiceReferenceSet<Library>(LIBRARY);
+
+    protected void activate(ComponentContext context, Map<String, Object> props) {
+        librarySet.activate(context);
+        for (ServiceReference<Library> ref : librarySet.references()) {
+            registerLibraryListener(context.getBundleContext(), (String) ref.getProperty("id"));
+        }
+    }
+
+    protected void deactivate(ComponentContext context) {
+        librarySet.deactivate(context);
```

Comment:
Do you need to "unregister" the Library listener on deactivate?  It all goes away maybe when deactivate is done?